private/buf/cmd/buf/command/dep/internal/internal.go

```diff
@@ -114,6 +120,34 @@ func Prune(
 	if err != nil {
 		return err
 	}
+	// If keepUnused is set to true, we add the unused declared dependencies to depModuleKeys.
+	if keepUnused {
+		moduleKeyProvider, err := bufcli.NewModuleKeyProvider(container)
+		if err != nil {
+			return err
+		}
+		unusedModuleKeys, err := moduleKeyProvider.GetModuleKeysForModuleRefs(
+			ctx,
+			unusedModuleRefs,
+			workspaceDepManager.BufLockFileDigestType(),
+		)
+		if err != nil {
+			return err
+		}
+		foundDepModuleKeys := copy(make([]bufmodule.ModuleKey, len(depModuleKeys)), depModuleKeys)
+		depModuleKeys = append(depModuleKeys, unusedModuleKeys...)
```

Comment:
I don't think this is right. You need to use `GraphProvider` to pull in the full dependency graph including those references (i.e., `internal.ModuleKeysAndTransitiveDepModuleKeysForModuleRefs`, which is what `dep update` does anyway). Without this, as soon as the user goes to import a file from the dependency, if that file's transitive dependencies are not found, the module will fail to build and the user will have to run `dep update` again. If you are using GraphProvider, you need to coordinate with `depModuleKeys` as well -- it needs to be entire enclosure.

Question: why do we prune the lock file at all in `dep update`? I think we want to remove unused deps, but isn't the core issue that the user just pulled it and is about to use it?

Reply:
> if that file's transitive dependencies are not found, the module will fail to build and the user will have to run dep update again.

Yeah, hence top-level comment around how the transitive dependencies wouldn't be pulled in. This seems somewhat reasonable because you don't know which transitive dependencies you need until you actually use the dependency.

> Question: why do we prune the lock file at all in dep update? I think we want to remove unused deps, but isn't the core issue that the user just pulled it and is about to use it?

Yeah, but there are also dependencies that have been removed from configuration (`buf.yaml`) that are being pruned from `buf.lock` at the time of `dep update`.

Reply:
I would be frustrated as a user if I had to run `dep update` twice, especially if I'm in an IDE and the tooling is complaining.

> dependencies that have been removed from configuration

Aren't this already pruned earlier in `dep update`? `dep update` is basically `ModuleKeysAndTransitiveDepModuleKeysForModuleRefs` + write to lock file, which should remove entries from lock file that are stale.

Reply:
By running `Prune` again, we are verifying the build, which does have some implications on tamper-proofing. But perhaps running the full `Prune` is not the way to go, as you are correct, the first write should already capture that (there is a clarifying comment also, my mistake on this one ^^").

Reply:
I made a change where we do not run `Prune` at all in `dep update`, but instead just verify the build and log unused deps. This way, we can ensure the transitive dependencies of the unused deps are included, I think you might be right, this might be the more elegant user experience.